hazelcast/src/main/java/com/hazelcast/nio/tcp/handlermigration/HandlerBalancer.java

```diff
@@ -0,0 +1,179 @@
+        if (log.isFinestEnabled()) {
+            log.finest("Adding a read handler " + readHandler);
+        }
+        inCalculator.addHandler(readHandler);
+
+        WriteHandler writeHandler = ((TcpIpConnection) connection).getWriteHandler();
+        if (log.isFinestEnabled()) {
+            log.info("Adding a write handler " + writeHandler);
+        }
+        outCalculator.addHandler(writeHandler);
+    }
+
+    public void connectionRemoved(Connection connection) {
+        if (!(connection instanceof TcpIpConnection)) {
+            return;
+        }
+
+        ReadHandler readHandler = ((TcpIpConnection) connection).getReadHandler();
+        if (log.isFinestEnabled()) {
+            log.finest("Removing a read handler " + readHandler);
+        }
+        inCalculator.removeHandler(readHandler);
+
+        WriteHandler writeHandler = ((TcpIpConnection) connection).getWriteHandler();
+        if (log.isFinestEnabled()) {
+            log.finest("Removing a write handler " + readHandler);
+        }
+        outCalculator.removeHandler(writeHandler);
+    }
+
+    @Override
+    public void run() {
+        try {
+            while (isActive) {
+                log.finest("Starting selector thread handler balance check");
+                checkReadHandlers();
+                checkWriteHandlers();
+                TimeUnit.SECONDS.sleep(migrationIntervalSeconds);
+            }
+        } catch (InterruptedException e) {
+            //this thread is about to exit, no reason restoring the interrupt flag
+            EmptyStatement.ignore(e);
+        }
+    }
+
+    public void stop() {
+        isActive = false;
+    }
+
+    public void start() {
+        Thread migratorThread = new Thread(this);
+        migratorThread.setDaemon(true);
+        migratorThread.start();
+    }
+
+    private void checkWriteHandlers() {
+        scheduleMigrationIfNeeded(outCalculator);
+    }
+
+    private void checkReadHandlers() {
+        scheduleMigrationIfNeeded(inCalculator);
+    }
+
+    public void scheduleMigrationIfNeeded(IOSelectorLoadCalculator stateCalculator) {
+        BalancerState balancerState = stateCalculator.getState();
+        if (strategy.shouldAttemptToScheduleMigration(balancerState)) {
+            tryToMigrate(balancerState);
+        } else {
+            if (log.isFinestEnabled()) {
+                long min = balancerState.minimumEvents;
+                long max = balancerState.maximumEvents;
+                log.finest("No imbalance has been detected. Max. events: " + max + " Min events: " + min + ".");
+            }
+        }
+    }
+
+    private boolean shouldStart(InSelectorImpl[] inSelectors, OutSelectorImpl[] outSelectors) {
+        return inSelectors.length > 1 || outSelectors.length > 1;
+    }
+
+    private void tryToMigrate(BalancerState balancerState) {
+        MigratableHandler handler = strategy.findHandlerToMigrate(balancerState);
+        if (handler != null) {
```

Comment:
I would go for

```
if(handler == null){
   log...
   return;
}

the cool stuff
```

This way you jump out of a method as soon as you figure out there is no point in staying. It also reduces cyclomatic complexity.